Region inference: Use outlives-static constraints in constraint search

amandasystems opened the pull request:
Revise the extra `r: 'static` constraints added upon universe issues to add an explanation, and use that explanation during constraint blame search. This greatly simplifies the region inference logic, which now does not need to reverse-engineer the event that caused a region to outlive `'static`.

This cosmetically changes the output of two UI tests. I blessed them i separate commits with separate motivations, but that can of course be squashed as desired. We probably want that.

The PR was extracted out of rust-lang/rust#130227 and consists of one-third of its functional payload.

r? lcnr

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15272511494/job/42951276557) [(plain)](https://github.com/rust-lang/rust/commit/9a1faceec6d273a9c420cbdfc4d4baf78b237e24/checks/42951276557/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#2 transferring dockerfile: 2.62kB done
#2 DONE 0.0s

#3 [internal] load metadata for docker.io/library/ubuntu:22.04
#3 ERROR: failed to copy: httpReadSeeker: failed open: unexpected status code https://registry-1.docker.io/v2/library/ubuntu/manifests/sha256:67cadaff1dca187079fce41360d5a7eb6f7dcd3745e53c79ad5efd8563118240: 429 Too Many Requests - Server message: toomanyrequests: You have reached your unauthenticated pull rate limit. https://www.docker.com/increase-rate-limit
------
 > [internal] load metadata for docker.io/library/ubuntu:22.04:
------
Dockerfile:1
--------------------
   1 | >>> FROM ubuntu:22.04
   2 |     
   3 |     ARG DEBIAN_FRONTEND=noninteractive
--------------------
ERROR: failed to solve: ubuntu:22.04: failed to resolve source metadata for docker.io/library/ubuntu:22.04: failed to copy: httpReadSeeker: failed open: unexpected status code https://registry-1.docker.io/v2/library/ubuntu/manifests/sha256:67cadaff1dca187079fce41360d5a7eb6f7dcd3745e53c79ad5efd8563118240: 429 Too Many Requests - Server message: toomanyrequests: You have reached your unauthenticated pull rate limit. https://www.docker.com/increase-rate-limit
Command failed. Attempt 2/5:
#0 building with "xenodochial_swartz" instance using docker-container driver

#1 [internal] load build definition from Dockerfile
#1 transferring dockerfile: 2.62kB done
#1 DONE 0.0s

#2 [internal] load metadata for docker.io/library/ubuntu:22.04
#2 ERROR: failed to copy: httpReadSeeker: failed open: unexpected status code https://registry-1.docker.io/v2/library/ubuntu/manifests/sha256:67cadaff1dca187079fce41360d5a7eb6f7dcd3745e53c79ad5efd8563118240: 429 Too Many Requests - Server message: toomanyrequests: You have reached your unauthenticated pull rate limit. https://www.docker.com/increase-rate-limit
------
 > [internal] load metadata for docker.io/library/ubuntu:22.04:
------
Dockerfile:1
--------------------
   1 | >>> FROM ubuntu:22.04
   2 |     
   3 |     ARG DEBIAN_FRONTEND=noninteractive
--------------------
ERROR: failed to solve: ubuntu:22.04: failed to resolve source metadata for docker.io/library/ubuntu:22.04: failed to copy: httpReadSeeker: failed open: unexpected status code https://registry-1.docker.io/v2/library/ubuntu/manifests/sha256:67cadaff1dca187079fce41360d5a7eb6f7dcd3745e53c79ad5efd8563118240: 429 Too Many Requests - Server message: toomanyrequests: You have reached your unauthenticated pull rate limit. https://www.docker.com/increase-rate-limit
Command failed. Attempt 3/5:
#0 building with "xenodochial_swartz" instance using docker-container driver

#1 [internal] load build definition from Dockerfile
#1 transferring dockerfile: 2.62kB done
#1 DONE 0.0s

#2 [internal] load metadata for docker.io/library/ubuntu:22.04
#2 ERROR: failed to copy: httpReadSeeker: failed open: unexpected status code https://registry-1.docker.io/v2/library/ubuntu/manifests/sha256:67cadaff1dca187079fce41360d5a7eb6f7dcd3745e53c79ad5efd8563118240: 429 Too Many Requests - Server message: toomanyrequests: You have reached your unauthenticated pull rate limit. https://www.docker.com/increase-rate-limit
------
 > [internal] load metadata for docker.io/library/ubuntu:22.04:
------
Dockerfile:1
--------------------
   1 | >>> FROM ubuntu:22.04
   2 |     
   3 |     ARG DEBIAN_FRONTEND=noninteractive
--------------------
ERROR: failed to solve: ubuntu:22.04: failed to resolve source metadata for docker.io/library/ubuntu:22.04: failed to copy: httpReadSeeker: failed open: unexpected status code https://registry-1.docker.io/v2/library/ubuntu/manifests/sha256:67cadaff1dca187079fce41360d5a7eb6f7dcd3745e53c79ad5efd8563118240: 429 Too Many Requests - Server message: toomanyrequests: You have reached your unauthenticated pull rate limit. https://www.docker.com/increase-rate-limit
Command failed. Attempt 4/5:
#0 building with "xenodochial_swartz" instance using docker-container driver

#1 [internal] load build definition from Dockerfile
#1 transferring dockerfile: 2.62kB done
#1 DONE 0.0s

#2 [internal] load metadata for docker.io/library/ubuntu:22.04
#2 ERROR: failed to copy: httpReadSeeker: failed open: unexpected status code https://registry-1.docker.io/v2/library/ubuntu/manifests/sha256:67cadaff1dca187079fce41360d5a7eb6f7dcd3745e53c79ad5efd8563118240: 429 Too Many Requests - Server message: toomanyrequests: You have reached your unauthenticated pull rate limit. https://www.docker.com/increase-rate-limit
------
 > [internal] load metadata for docker.io/library/ubuntu:22.04:
------
Dockerfile:1
--------------------
   1 | >>> FROM ubuntu:22.04
   2 |     
   3 |     ARG DEBIAN_FRONTEND=noninteractive
--------------------
ERROR: failed to solve: ubuntu:22.04: failed to resolve source metadata for docker.io/library/ubuntu:22.04: failed to copy: httpReadSeeker: failed open: unexpected status code https://registry-1.docker.io/v2/library/ubuntu/manifests/sha256:67cadaff1dca187079fce41360d5a7eb6f7dcd3745e53c79ad5efd8563118240: 429 Too Many Requests - Server message: toomanyrequests: You have reached your unauthenticated pull rate limit. https://www.docker.com/increase-rate-limit
Command failed. Attempt 5/5:
#0 building with "xenodochial_swartz" instance using docker-container driver

#1 [internal] load build definition from Dockerfile
#1 transferring dockerfile: 2.62kB done
#1 DONE 0.0s

#2 [internal] load metadata for docker.io/library/ubuntu:22.04
#2 ERROR: failed to copy: httpReadSeeker: failed open: unexpected status code https://registry-1.docker.io/v2/library/ubuntu/manifests/sha256:67cadaff1dca187079fce41360d5a7eb6f7dcd3745e53c79ad5efd8563118240: 429 Too Many Requests - Server message: toomanyrequests: You have reached your unauthenticated pull rate limit. https://www.docker.com/increase-rate-limit
------
 > [internal] load metadata for docker.io/library/ubuntu:22.04:
------
Dockerfile:1
--------------------
   1 | >>> FROM ubuntu:22.04
   2 |     
   3 |     ARG DEBIAN_FRONTEND=noninteractive
--------------------
ERROR: failed to solve: ubuntu:22.04: failed to resolve source metadata for docker.io/library/ubuntu:22.04: failed to copy: httpReadSeeker: failed open: unexpected status code https://registry-1.docker.io/v2/library/ubuntu/manifests/sha256:67cadaff1dca187079fce41360d5a7eb6f7dcd3745e53c79ad5efd8563118240: 429 Too Many Requests - Server message: toomanyrequests: You have reached your unauthenticated pull rate limit. https://www.docker.com/increase-rate-limit
The command has failed after 5 attempts.
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15306187270/job/43059485861) [(plain)](https://github.com/rust-lang/rust/commit/6539053fb34368678864c7e96505ded7cfaacbf7/checks/43059485861/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
No error code explanation was removed!
+ reuse --include-submodules lint
# MISSING COPYRIGHT AND LICENSING INFORMATION

The following files have no copyright and licensing information:
* ../elmininate_placeholders.rs


# SUMMARY

* Bad licenses: 0
---


# RECOMMENDATIONS

* Fix missing copyright/licensing information: For one or more files, the tool
  cannot find copyright and/or licensing information. You typically do this by
  adding 'SPDX-FileCopyrightText' and 'SPDX-License-Identifier' tags to each
  file. The tutorial explains additional ways to do this:
  <https://reuse.software/tutorial/>

  local time: Wed May 28 17:34:17 UTC 2025
  network time: Wed, 28 May 2025 17:34:17 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

lcnr commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang/rust/actions/runs/15533835962/job/43737552897)
Build commit: 0d3d48082efe47a20c953415945cc721105e5f85 (`0d3d48082efe47a20c953415945cc721105e5f85`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang/rust/actions/runs/15533835962/job/43737552897"},"merge_sha":"0d3d48082efe47a20c953415945cc721105e5f85"} -->

lcnr commented on compiler/rustc_borrowck/src/handle_placeholders.rs:
```diff
@@ -64,80 +68,162 @@ impl scc::Annotations<RegionVid> for SccAnnotations<'_, '_, RegionTracker> {
 
     /// Determine if the tracked universes of the two SCCs are compatible.
     pub(crate) fn universe_compatible_with(&self, other: Self) -> bool {
         self.max_nameable_universe().can_name(other.max_nameable_universe())
```
```suggestion
        // HACK: We first check whether we can name the highest existential universe
        // of `other`. This only exists to avoid errors in case that scc already
        // has a placeholder error itself.
        self.max_nameable_universe().can_name(other.max_nameable_universe())
```

lcnr commented on compiler/rustc_borrowck/src/handle_placeholders.rs:
```diff
@@ -321,28 +410,103 @@ fn rewrite_placeholder_outlives<'tcx>(
+            annotation.representative
+        );
+        // We only add one `r: 'static` constraint per SCC, where `r` is the SCC representative.
+        // That constraint is annotated with some outlives relation `tries: unnameable` where
+        // `unnameable` is unnameable from `tries` and there is a path in the constraint
```
what is `tries` here? just some region var? `'lt: 'unnameable` maybe :thinking:

amandasystems replied:
Makes sense, I'll make it so.

lcnr commented on compiler/rustc_borrowck/src/handle_placeholders.rs:
```diff
@@ -321,28 +410,103 @@ fn rewrite_placeholder_outlives<'tcx>(
+        let blame_to = if annotation.representative.rvid() == max_u_rvid {
+            // The SCC's representative is not nameable from some region
+            // that ends up in the SCC.
+            let small_universed_rvid = find_region(
```
why can we not use the region from `max_nameable_universe` here?

amandasystems replied:
Here's the an output from the UI tests attached when doing that!

```
$ ./x test ui --keep-stage 1 > simpler-choice-blame-ui-tests.log
```
[simpler-choice-blame-ui-tests.log](https://github.com/user-attachments/files/21433258/simpler-choice-blame-ui-tests.log)

There's a bunch of squiggly lines getting longer, which I think is mostly harmless:
- tests/ui/higher-ranked/higher-ranked-lifetime-error.rs: I'd say fully harmless, a slightly longer squiggly line
- tests/ui/higher-ranked/trait-bounds/due-to-where-clause.rs same as above
-  tests/ui/mismatched_types/hr-projection-mismatch.rs#current same as above

Then there is a duplicate of a diagnostic due to a longer span which we may or may not care about, in `tests/ui/associated-inherent-types/issue-111404-1.rs`. This should be very fixable in the error reporting phase, but that part of the code is absolutely incomprehensible (though I've tried), and I've simply given up on it.

I remember this being much, much worse! Maybe this is acceptable?

amandasystems replied:
If you think this is ok, say the word and I'll dump like half the file to get rid of this extra logic!

lcnr replied:
This is okay to me and I would prefer getting rid of this extra logic. Though I could imagine us improving these diagnostics with at least one hacky way. Instead of using the first `max_universe`, we could do fun stuff with it. If we're able to pass in the regionctxt/region info into `merge_sccs` we can decide the best region based on its origin/span.

What happens if `pick_min_max_universe` just uses the region with the lowest VID if universes are equal?

lcnr replied:
:thinking: that seems desirable, except for the change to `tests/ui/generic-associated-types/bugs/hrtb-implied-2.rs`

We emit this limitation error if the constraint path contains some constraint introduced by a `whatever: 'placeholder` constraint. I guess we end up with a different constraint path which never has such a a `Predicate` constraint?

I guess we'd want to look at the constraint path we get with this change for the region error in this test and check why it's not what we expect?

amandasystems replied:
Here's the blame picks:

```
0ms DEBUG rustc_borrowck::region_infer constraint ('?5: '?19) due to Single(bb1[6]) (None) (BoringNoLocation) category: BoringNoLocation, interest: 6
   0ms DEBUG rustc_borrowck::region_infer constraint ('?10: '?5) due to Single(bb1[6]) (Invariant { ty: &'?13 mut Eat<&'?5 mut I/#0, F/#1>, param_index: 0 }) (CallArgument(Some(FnDef(DefId(2:9335 ~ core[4df0]::iter::traits::iterator::Iterator::next), [Eat<&'{erased} mut I/#0, F/#1>])))) category: CallArgument(Some(FnDef(DefId(2:9335 ~ core[4df0]::iter::traits::iterator::Iterator::next), [Eat<&'{erased} mut I/#0, F/#1>]))), interest: 2
   0ms DEBUG rustc_borrowck::region_infer constraint ('?7: '?10) due to Single(bb1[5]) (Invariant { ty: &'?9 mut Eat<&'?10 mut I/#0, F/#1>, param_index: 0 }) (Boring) category: Boring, interest: 5
   0ms DEBUG rustc_borrowck::region_infer constraint ('?3: '?7) due to Single(bb0[5]) (None) (Assignment) category: Assignment, interest: 2
   0ms DEBUG rustc_borrowck::region_infer constraint ('?8: '?3) due to Single(bb0[5]) (None) (Boring) category: Boring, interest: 5
   0ms DEBUG rustc_borrowck::region_infer constraint ('?6: '?8) due to Single(bb0[2]) (None) (Usage) category: Usage, interest: 2
   0ms DEBUG rustc_borrowck::region_infer constraint ('?1: '?6) due to All(tests/ui/generic-associated-types/bugs/hrtb-implied-2.rs:13:33: 13:37 (#0)) (None) (BoringNoLocation) category: BoringNoLocation, interest: 6
   0ms DEBUG rustc_borrowck::region_infer best_choice=5, blame_source=true
```

lcnr replied:
this is a lot of text and it's late in the evening, but a quick question, what's the constraint from `'?5: '?19`? I guess it's not a predicate constraint, but what other reason exists for us to add an outlives placeholder bound here?

amandasystems replied:
> It looks like it's a BoringNoLocation, but why xd.

*heavy drag on cigarette* I don’t know man. I don’t make the constraint graph, I just walk it.

> The important change to avoid the diagnostics regression is to choose the placeholder with the smallest vid, ignoring its universe once both placeholder are no longer nameable? It feels annoyingly arbitrary 😅 

it is a bit hacky. What it actually does is kind of approximate picking an issue early in the constraint graph, and it turns out smaller rvids appear earlier. This is a pretty extreme case though in that the distance to all the errors is the same and the only thing guaranteeing the behaviour is order of insertion in successors for ?5 in the constraint graph. Which uh also feels arbitrary.

amandasystems replied:
The only thing that wouldn’t be arbitrary would be to, as my wife suggested, track literally all of them. That would make the whole SCC annotation situation absolutely terrible because they’re no longer copy and I can’t imagine that wouldn’t have performance implications. 

Another option might be to add an outlives static constraint per unnameable etc etc, which would turn the constraint graph into spaghetti.

lcnr replied:
> heavy drag on cigarette I don’t know man. I don’t make the constraint graph, I just walk it.

would be good to know :grin: it implies we somehow end up relating something to a placeholder from a higher universe for a boring reason? As in, even without the hr region error due to the predicate we'd have failed to compile this snippet

---

I feel like mir borrowck generally has to handle the following for when encountering an error:

- there may be multiple paths causing the bad outlives (idk about terminology)
- each path has multiple steps

I generally feel like we want to report the *most boring and shortest* path and from that path, the *most exciting* step.

A step is boring if it exists due to reasons the user can't do much about, e.g. for types to be wf

A step is *exciting*:tm: if it corresponds to something which is actively happening in the source, some where-bound, some assignment. The user is likely to be change a specific part of their code to avoid emitting this constraint.

So in this example, if we get an outlives `'static` constraint both because its related to a placeholder for "boring" reasons, and also due to a predicate, we want to treat it as if its due to the boring reason, as fixing the predicate issue likely won't handle the boring constraint.

So yeah, I think it's actually good that we use this boring constraint instead of the predicate constraint here :>

lcnr replied:
I think given that the outlives 19 exist with a `BoringNoLocation`, the predicate is not to blame here. Now, I don't see why that boring region outlives placeholder exists, but imo as long as that outlives exists, we shouldn't add a note stating that it's due to a predicate, because there's a separate "non predicate" reason

I would love for u to look into why that other constraint exists.

> Constraint blame search also optionally tells in its result if search was redirected

hmm, that is another question :thinking: if we have `'n: 'placeholder` as a cause of a `'m: 'static` path, why do we compute the `'n: 'placeholder` path instead of the full `'m: 'placeholder` path (going through n).

Sorry for giving u a bunch of vague "pls look at X" reviews here btw. I think it's really good that we/you are cleaning this up but I want to make sure we're slowly moving towards something desirable while doing so

lcnr replied:
from placeholder to an existential should error, shouldn't it?

we only add `outlives-static` whenever we've got "scc_which_cannot_name reaches(transitively outlives) placeholder" constraints :thinking: 

My question above was, why does https://github.com/rust-lang/rust/blob/b03635d15ff684f3408110a3d7aff8be656143f5/compiler/rustc_borrowck/src/region_infer/mod.rs#L1796-L1810 check `constraint_path_to(lt, ..)` instead of `constraint_path_to(from_region, ..)`

amandasystems replied:
> I would love for u to look into why that other constraint exists.

This is my next order of business, because if I can show that then we can bless the note-less output and have an Even Simpler Implementation™

amandasystems replied:
Now I know what's happened, and I can explain why. I introduced a bug in SCC metadata merging that essentially messed it all up. It took the whole day to find it.

The logic currently on master actually introduces edges to 'static more often than you'd think, and this is the logic that helps detect placeholder-outlives-existential, which is why I accidentally turned it off.

What happens is this: an SCC `S` has representative region `r`. `r` reaches a region `'e` that can't name it, and so has its universe lowered to the `e`'s universe. This causes `r` to be marked by my code as unnameable from `S`, which adds an edge from `r` (as `S`'s representative) to `'static`. When selecting how to annotate the edge, I would prefer to have it look for an unnameable placeholder. Unfortunately, that's `r`. In the original code, which you commented on, this instead picks the smallest-univers'd region to blame (e.g. `'e`) for the constraint edge to `'static`, and starts a new search from `'r` to `'e` via the annotation.

In that case, we produce an explanation `‘representative: ‘low-u`, which is nice because it always leads to a path since we now have guarantees that the representative never appears as both start and end of the path. If there is an unnameable placeholder, it’s not the representative.

So, returning to your case:

If we have `'n: 'placeholder` as a cause of a `'m: 'static` path, we can end up with the case where `m = ‘placeholder` (idk it’s one half of a cycle or something). However, it turns out this is neatly handled by just not doing the redirect search in that case, and we find some interesting blame constraints along that path anyway.

Looking into the case of the missing note now!

amandasystems replied:
Nice, thanks!

lcnr commented on compiler/rustc_middle/src/mir/query.rs:
```diff
@@ -149,8 +149,17 @@ pub enum ConstraintCategory<'tcx> {
+    /// An internal constraint derived from an illegal placeholder relation
+    /// to this region. The arguments are a source -> drain of a path
+    /// that caused the problem, used when reporting errors.
+    IllegalPlaceholder(
```
I don't quite like this name as it seems more 'serious' than it should be, maybe `OutlivesUnnameablePlaceholder`?

amandasystems replied:
I've had the same feeling but couldn't come up with anything. That's better, thanks!

lcnr commented on compiler/rustc_borrowck/src/region_infer/mod.rs:
```diff
@@ -1782,42 +1730,48 @@ impl<'tcx> RegionInferenceContext<'tcx> {
         None
     }
 
+    /// Find a path of outlives constraints from `from` to `to`,
+    /// taking placeholder blame constraints into account, e.g.
+    /// if there is a relationship where `r1` reaches `r2` and
+    /// r2 has a larger universe or if r1 and r2 both come from
+    /// placeholder regions.
+    ///
+    /// Returns the path. It panics if there is no such path.
+    fn path_to_modulo_placeholders(
```
This function confuses me :sweat_smile: 

the idea is that if we've got `'from: 'to` which does not hold and to is actually just `'static`, then we want to report `'from: 'placeholder` if that's what caused the error.

More specifically, we don#t even care about `'from: 'placeholder`, we care about `'existential_which_cannot_name: 'placeholder` as the most specific error.

So it's less `path_to_modulo_placeholders` or "taking placeholder blame constraints into account", and more 

> Find a path of outlives constraints from `from` to `to`. In case this path exists due to an unnameable placeholder, returns the path from the existential region to the placeholder which it cannot name instead.
> It panics if there doesn't exist a path from `from` to `to`.

maybe something like `fn constraint_or_outlives_unnameable_placeholder_path` would be slightly clearer :thinking: it feels difficult to get a good name.

Alternatively, this method feels quite specific, maybe inline it into `best_blame_constraint`? :thinking:

amandasystems replied:
I'll check if the upstream changes in the other PRs needs the logic repeated somewhere, otherwise it's probably better to fold it into `best_blame_constraint` as you say!

lcnr commented on compiler/rustc_borrowck/src/region_infer/graphviz.rs:
```diff
@@ -12,6 +12,9 @@ use rustc_middle::ty::UniverseIndex;
+    if let ConstraintCategory::OutlivesUnnameablePlaceholder(from, to) = constraint.category {
+        return format!("b/c {from:?}: {to:?}");
+    }
     match constraint.locations {
```
style nit, use an else and don't explicitly return

lcnr commented on compiler/rustc_borrowck/src/region_infer/mod.rs:
```diff
@@ -1787,37 +1735,11 @@ impl<'tcx> RegionInferenceContext<'tcx> {
         trace!(universe = ?self.max_nameable_universe(self.constraint_sccs.scc(fr1)));
-        self.find_constraint_paths_between_regions(fr1, |r| {
+        self.constraint_path_to(fr1, |r| {
             // First look for some `r` such that `fr1: r` and `r` is live at `location`
```
```suggestion
```

lcnr commented on compiler/rustc_borrowck/src/region_infer/mod.rs:
```diff
@@ -1861,22 +1783,32 @@ impl<'tcx> RegionInferenceContext<'tcx> {
+
+        // If we are passing through a constraint added because `'lt: 'unnameable`,
+        // where cannot name `'unnameable`, redirect search towards `'unnameable`.
+        let path = if let Some((lt, unnameable)) = path.iter().find_map(|c| {
+            if let ConstraintCategory::OutlivesUnnameablePlaceholder(lt, unnameable) = c.category {
+                Some((lt, unnameable))
+            } else {
+                None
+            }
+        }) {
```
```suggestion
        let due_to_placeholder_outlives = path.iter().find_map(|c| {
            if let ConstraintCategory::OutlivesUnnameablePlaceholder(lt, unnameable) = c.category {
                Some((lt, unnameable))
            } else {
                None
            }
        });
        let path = if let Some((lt, unnameable)) = due_to_placeholder_outlives {
```

bors commented:
:umbrella: The latest upstream changes (presumably #144469) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

amandasystems committed:
Region inference: Use outlives-static constraints in constraint search

Revise the extra `r: 'static` constraints added upon universe issues
to add an explanation, and use that explanation during constraint blame
search. This greatly simplifies the region inference logic, which
now does not need to reverse-engineer the event that caused a region
to outlive 'static.

amandasystems commented:
We may also want a new perf run since I have changed the code that runs during SCC construction which might affect performance (positively, I hope)

lcnr commented:
@bors try @rust-timer queue

lcnr commented on compiler/rustc_borrowck/src/region_infer/mod.rs:
```diff
@@ -1864,22 +1784,37 @@ impl<'tcx> RegionInferenceContext<'tcx> {
+            }
+        });
+
+        // Edge case: it's possible that `'from_region` is an unnameable placeholder.
```
what exactly is this edge case and when does it occur, also

> // This the `false` argument is what prevents circular reasoning here!

lcnr replied:
> I meant that there is a risk of this being infinitely recursive if this passes through the same outlives static constraint again.

do you have a specific example in mind?

so this matters if we have `'n: 'm` in `best_blame_constraint`.

The initial path is `'n -> 'a -> -OutlivesUnnameablePlaceholder('p)-> 'static -> 'm`

We then lookup `'n: 'p` which should give us `'n -> 'a -> 'b -> 'p`.

The worry is that there's some path `'n -> 'a -> 'static -> 'p`. i.e. cases where `'a` is already required to outlive `'static`? I feel like such a path is kind of sus as really, we're already getting an error even without the placeholder, so the `OutlivesUnnameablePlaceholder` constraint isn't actually required for the original path :thinking:  

I don't see how this could be infinitely recursive though :thinking: we only visit regions with  `Trace::NotVisited`, so this should always terminate, shouldn't it?

> // [This|the] `false` argument is what prevents circular reasoning here!

This is what I wanted to state. "This the `false` argument" is a just wrong :>

amandasystems replied:
> The worry is that there's some path 'n -> 'a -> 'static -> 'p. i.e. cases where 'a is already required to outlive 'static? I feel like such a path is kind of sus as really, we're already getting an error even without the placeholder, so the OutlivesUnnameablePlaceholder constraint isn't actually required for the original path 🤔

You're right! This doesn't happen in the tests, but it could be an issue I guess, and it slipped my mind. I've fixed it in a version that passes the test suite equally well (soon to be pushed!). The uh messed up comment is also fixed.

I was confused about the infinite recursion because for a very long time in my parallel universe all graph searches would redirect on these new constraints, and in that case it _would_ be an infinite loop. I just forgot that was no longer the case. Sorry!

I'm very tempted to restructure graph search as a builder that produces a configuration `GraphSearch` so you can go `GraphSearch::between(r1, r2).ignoring_extra_constraints().search(region_graph, member_constraints)` but I'll contain myself.

rust-timer commented:
Queued 834fbf1e2aa01b56f03cb70cc1f290e146189916 with parent 8365fcb2b840c95eeb0bc377af8bd498fad22245, future [comparison URL](https://perf.rust-lang.org/compare.html?start=8365fcb2b840c95eeb0bc377af8bd498fad22245&end=834fbf1e2aa01b56f03cb70cc1f290e146189916).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.1 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

bors commented:
:umbrella: The latest upstream changes (presumably #145244) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

amandasystems commented on compiler/rustc_borrowck/src/handle_placeholders.rs:
```diff
@@ -314,28 +378,52 @@ fn rewrite_placeholder_outlives<'tcx>(
+
+        added_constraints = true;
+        outlives_constraints.push(OutlivesConstraint {
+            sup: annotation.representative.rvid(),
```
I think that interferes with the tracing code/search redirect I currently have but you’re right that would make more sense. I’ll see if it’s easily tweakable.

amandasystems replied:
(I suspect this is also mitigated when placeholder errors are handled separately later and the logic of this part _only_ does one thing, but I worry there might be residual bits of the error explanation code that assumes that something like this structure exists because it was how the do-the-logic-during-propagation code used to work)

lcnr commented on compiler/rustc_borrowck/src/region_infer/mod.rs:
```diff
@@ -2025,7 +1970,7 @@ impl<'tcx> RegionInferenceContext<'tcx> {
                 ConstraintCategory::BoringNoLocation => 6,
                 // Do not blame internal constraints.
-                ConstraintCategory::IllegalUniverse => 7,
+                ConstraintCategory::OutlivesUnnameablePlaceholder(_) => 7,
```
change this to `unreachable`? :grin:

lcnr commented:
needs rebase :>

amandasystems committed:
Simplify SCC annotations somewhat

lcnr commented:
@bors retry

Kobzol commented:
The post merge perf. run was kind of reversed from the pre merge perf. run. Do you think it's worth it to investigate if the regression could be clawed back?

amandasystems commented:
> The post merge perf. run was kind of reversed from the pre merge perf. run. Do you think it's worth it to investigate if the regression could be clawed back?

I’m not sure how there could be actual differences between them and not just noise, but I think some of the real regressions come from actually computing slightly more stuff during SCC construction and getting around that is maybe not possible at all with this approach. 

The thing I *do* have a sketch for is how to avoid recomputing SCCs after merging, but it’ll be a while until I have the time to work on it due to baby.

Hopefully we can get some perf back when we reach the end goal of not tracking placeholders at all though!

Kobzol commented:
Ok, sounds good enough to me!

@rustbot label: +perf-regression-triaged
